### PR TITLE
[file upload plugin] allow for async file url resolution

### PR DESCRIPTION
**Pull request by vybu**
We want to be able to resolve file url on demand, this way we can handle private file downloading.

On file upload we will not save file url to content state, but some other metadata which will be used by url resolver to resolve it on demand (while also checking if user has necessary permissions)

### CODE REVIEW

File: packages/plugin-file-upload/src/file-upload-viewer.jsx [hunk 1]
```diff
@@ -44,37 +49,87 @@ class FileUploadViewer extends PureComponent {
...
 
+  renderFileUrlResolver() {
+    const { error, componentData, settings } = this.props;
+
+    if (error) {
+      return null;
+    }
+
+    const resolveFileUrl = () => {
+      settings.resolveFileUrl(componentData).then(resolveFileUrl => {
```
Comment (lxgreen) [inline packages/plugin-file-upload/src/file-upload-viewer.jsx]: requires 'settings' and 'resolveFileUrl' validation before call

File: packages/plugin-file-upload/src/file-upload-viewer.jsx [hunk 2]
```diff
@@ -84,10 +139,12 @@ FileUploadViewer.propTypes = {
   isLoading: PropTypes.bool.isRequired,
   componentData: PropTypes.object.isRequired,
   error: PropTypes.string,
+  settings: PropTypes.object.isRequired,
```
Comment (lxgreen) [inline packages/plugin-file-upload/src/file-upload-viewer.jsx]: optional

File: packages/plugin-file-upload/src/file-upload-viewer.jsx [hunk 3]
```diff
@@ -44,37 +49,87 @@ class FileUploadViewer extends PureComponent {
...
+  }
+
+  renderAutoDownloadIframe() {
+    const withFileUrlResolver = this.props.settings.resolveFileUrl;
+
+    if (!withFileUrlResolver) {
+      return null;
+    }
+
+    return <iframe id={this.fileDownloadIframeId} style={{ display: 'none' }} title="file" />;
```
Comment (lxgreen) [inline packages/plugin-file-upload/src/file-upload-viewer.jsx]: please consider using ref instead id

Comment (vybu) [reply packages/plugin-file-upload/src/file-upload-viewer.jsx]: I tried but for some reason download wouldn't trigger :thinking:

**Pull request merged**